File: content_tool/agents/topic_gen.py
# ruff: noqa: RUF001
"""Topic-generation agent.

One Gemini call: take a research brief and produce a list of pillar-topic
candidates with focus keywords. No DB writes — the caller (topic-expansion
subgraph in Task 3) persists the rows.
"""

from content_tool import prompts_store
from content_tool.gemini.client import GeminiClient
from content_tool.models.topic_batch import TopicGenInput, TopicGenOutput


def _format_list_block(items: list[str]) -> str:
    if not items:
        return "（無）"
    return "\n".join(f"- {item}" for item in items)


async def build_system_prompt() -> str:
    return await prompts_store.get_assembled_standalone("topic_gen")


def build_user_prompt(input_: TopicGenInput) -> str:
    return (
        "請根據以下研究設定產出結果。\n\n"
        f"研究主題：{input_.research_theme}\n"
        f"目標受眾：{input_.target_audience}\n"
        f"主題數量：{input_.topic_count}\n"
        f"每個主題關鍵字數量：{input_.keywords_per_topic}\n\n"
        f"必須涵蓋範疇：\n{_format_list_block(input_.must_cover)}\n\n"
        f"避免主題：\n{_format_list_block(input_.must_avoid)}\n\n"
        f"額外偏重方向：\n{input_.priority_focus or '（無）'}\n\n"
        f"補充要求：\n{input_.notes or '（無）'}\n"
    )


async def run_topic_gen(
    *,
    gemini: GeminiClient,
    input: TopicGenInput,
) -> TopicGenOutput:
    """Single Gemini call. Returns validated topic candidates."""
    system_prompt = await build_system_prompt()
    user_prompt = build_user_prompt(input)
    result = await gemini.generate(
        agent="topic_gen",
        system_prompt=system_prompt,
        user_prompt=user_prompt,
        response_schema=TopicGenOutput.model_json_schema(),
        tools=["googleSearch", "urlContext"],
    )
    return TopicGenOutput.model_validate(result.parsed)

File: deploy/cloudflare-workers/src/agents/topic_gen.ts
/**
 * Topic-generation agent — TypeScript port of `content_tool/agents/topic_gen.py`.
 *
 * One Gemini call: take a research brief and produce a list of pillar-topic
 * candidates with focus keywords. No DB writes — the caller (the
 * TopicExpansionWorkflow, assembled by the lead) persists the rows; this
 * function returns the parsed candidates plus token usage.
 *
 * Full-width punctuation (：、（無）) is intentional — the prompts are CJK style,
 * matching the Python `build_user_prompt` byte-for-byte.
 */

import type { Sql } from "postgres";
import { getAssembled } from "../prompts/store";
import type { GeminiClient, ThoughtCallback } from "../gemini/types";
import { TOPIC_GEN_SCHEMA, type TopicGenOutput } from "./topic_schemas";

// ---------------------------------------------------------------------------
// Input — mirrors Python TopicGenInput (content_tool/models/topic_batch.py)
// ---------------------------------------------------------------------------

export interface TopicGenInput {
  researchTheme: string;
  targetAudience: string;
  topicCount: number;
  keywordsPerTopic: number;
  mustCover: string[];
  mustAvoid: string[];
  priorityFocus?: string | null;
  notes?: string | null;
  /** The batch's voice (persona slug); resolves the prompt under it. */
  voiceSlug: string;
  onThought?: ThoughtCallback;
}

// ---------------------------------------------------------------------------
// Token usage result (mirrors OutlineTokens)
// ---------------------------------------------------------------------------

export interface TopicGenTokens {
  tokensIn: number;
  tokensOut: number;
  thinkingTokens: number;
  latencyMs: number;
}

// ---------------------------------------------------------------------------
// System prompt assembly — mirrors Python `build_system_prompt`.
// ---------------------------------------------------------------------------

async function buildSystemPrompt(sql: Sql, voiceSlug: string): Promise<string> {
  return getAssembled(sql, "topic_gen", voiceSlug);
}

// ---------------------------------------------------------------------------
// List block — mirrors Python `_format_list_block`.
// ---------------------------------------------------------------------------

function formatListBlock(items: string[]): string {
  if (items.length === 0) {
    return "（無）";
  }
  return items.map((item) => `- ${item}`).join("\n");
}

// ---------------------------------------------------------------------------
// User prompt — mirrors Python `build_user_prompt` exactly.
// ---------------------------------------------------------------------------

export function buildUserPrompt(opts: {
  researchTheme: string;
  targetAudience: string;
  topicCount: number;
  keywordsPerTopic: number;
  mustCover: string[];
  mustAvoid: string[];
  priorityFocus?: string | null;
  notes?: string | null;
}): string {
  return (
    "請根據以下研究設定產出結果。\n\n" +
    `研究主題：${opts.researchTheme}\n` +
    `目標受眾：${opts.targetAudience}\n` +
    `主題數量：${opts.topicCount}\n` +
    `每個主題關鍵字數量：${opts.keywordsPerTopic}\n\n` +
    `必須涵蓋範疇：\n${formatListBlock(opts.mustCover)}\n\n` +
    `避免主題：\n${formatListBlock(opts.mustAvoid)}\n\n` +
    `額外偏重方向：\n${opts.priorityFocus || "（無）"}\n\n` +
    `補充要求：\n${opts.notes || "（無）"}\n`
  );
}

// ---------------------------------------------------------------------------
// Main export
// ---------------------------------------------------------------------------

/**
 * Run the topic-gen agent. Assembles the system + user prompt, calls Gemini with
 * structured output and the googleSearch + urlContext grounding tools, and
 * returns the parsed candidates plus token usage. No DB writes.
 */
export async function runTopicGen(
  sql: Sql,
  gemini: GeminiClient,
  input: TopicGenInput,
): Promise<{ output: TopicGenOutput; tokens: TopicGenTokens }> {
  const systemPrompt = await buildSystemPrompt(sql, input.voiceSlug);
  const userPrompt = buildUserPrompt({
    researchTheme: input.researchTheme,
    targetAudience: input.targetAudience,
    topicCount: input.topicCount,
    keywordsPerTopic: input.keywordsPerTopic,
    mustCover: input.mustCover,
    mustAvoid: input.mustAvoid,
    priorityFocus: input.priorityFocus,
    notes: input.notes,
  });

  const result = await gemini.generate({
    agent: "topic_gen",
    systemPrompt,
    userPrompt,
    responseSchema: TOPIC_GEN_SCHEMA as Record<string, unknown>,
    tools: ["googleSearch", "urlContext"],
    onThought: input.onThought,
  });

  // Gemini returns a plain object matching TopicGenOutput's shape when
  // responseSchema is provided. Route through `unknown` since
  // `GeminiResult.parsed` is `Record<string, unknown>`.
  const output = result.parsed as unknown as TopicGenOutput;

  return {
    output,
    tokens: {
      tokensIn: result.tokensIn,
      tokensOut: result.tokensOut,
      thinkingTokens: result.thinkingTokens,
      latencyMs: result.latencyMs,
    },
  };
}
